src/coreclr/vm/gccover.cpp

```diff
@@ -1845,29 +1207,172 @@ void DoGcStress (PCONTEXT regs, NativeCodeVersion nativeCodeVersion)
+    _ASSERTE(!nativeCodeVersion.IsNull());
+    GCCoverageInfo *gcCover = nativeCodeVersion.GetGCCoverageInfo();
+    if (gcCover == nullptr)
+    {
+        return false;
+    }
+
+    PBYTE instrPtr = reinterpret_cast<PBYTE>(ip);
+
+    if (IsGcCoverageInterruptInstruction(instrPtr))
+    {
+        return true;
+    }
+
+    if (IsOriginalInstruction(instrPtr, gcCover, codeInfo.GetRelOffset()))
+    {
+        // Another thread may have already changed the code back to the original.
+        return true;
     }
+    return false;
+}
+
+// Remove the GcCoverage interrupt instruction, and restore the
+// original instruction. Only one instruction must be used,
+// because multiple threads can be executing the same code stream.
+
+void RemoveGcCoverageInterrupt(TADDR instrPtr, BYTE * savedInstrPtr, GCCoverageInfo* gcCover, DWORD offset)
+{
+    ExecutableWriterHolder<void> instrPtrWriterHolder((void*)instrPtr, 4);
+#ifdef TARGET_ARM
+    if (GetARMInstructionLength(savedInstrPtr) == 2)
+        *(WORD *)instrPtrWriterHolder.GetRW()  = *(WORD *)savedInstrPtr;
+    else
+        *(DWORD *)instrPtrWriterHolder.GetRW() = *(DWORD *)savedInstrPtr;
+#elif defined(TARGET_ARM64)
+    *(DWORD *)instrPtrWriterHolder.GetRW() = *(DWORD *)savedInstrPtr;
+#elif defined(TARGET_LOONGARCH64)
+    *(DWORD *)instrPtrWriterHolder.GetRW() = *(DWORD *)savedInstrPtr;
+#elif defined(TARGET_RISCV64)
+    *(DWORD *)instrPtrWriterHolder.GetRW() = *(DWORD *)savedInstrPtr;
+#else
+    *(BYTE *)instrPtrWriterHolder.GetRW() = *savedInstrPtr;
 #endif
 
-    if (!Thread::UseRedirectForGcStress())
+#ifdef TARGET_X86
+    // Epilog checking relies on precise control of when instrumentation for the  first prolog
+    // instruction is enabled or disabled. In particular, if a function has multiple epilogs, or
+    // the first execution of the function terminates via an exception, and subsequent completions
+    // do not, then the function may trigger a false stress fault if epilog checks are not disabled.
+    if (offset == 0)
     {
-        frame.Pop(pThread);
+        gcCover->doingEpilogChecks = false;
     }
+#endif // TARGET_X86
 
-    if (enableWhenDone)
+    FlushInstructionCache(GetCurrentProcess(), (LPCVOID)instrPtr, 4);
+}
+
+// A managed thread (T) can race with the GC as follows:
+// 1) Current thread T is in preemptive mode, GC happens, starts scanning T's stack.
+// 2) The Jitted code puts thread T to cooperative mode, as part of PInvoke epilog.
+// 3) Jitted code checks g_TrapReturningThreads, calls CORINFO_HELP_STOP_FOR_GC()
+// 4) If any of the code that does #3 is interruptible, we may hit GCStress trap and start
+//    another round of GCStress while in Cooperative mode.
+// 5) Now, thread T can modify the stack (ex: RedirectionFrame setup) while the GC thread is scanning it.
+//
+// This race is now mitigated below. Where we won't initiate a stress mode GC
+// for a thread in cooperative mode with an active ICF, if g_TrapReturningThreads is true.
```

Comment:
The original explanation for the g_TrapReturningThreads + active ICF scenario causing a race was somewhat questionable, so I initially removed the mitigation, but the race appears to be real. 

Corrected the explanation of how the race may happen and brought the mitigation back.